kong/router/atc_compat.lua

```diff
@@ -0,0 +1,468 @@
+    error("dst_ip must be a string", 2)
+  end
+  if dst_port and type(dst_port) ~= "number" then
+    error("dst_port must be a number", 2)
+  end
+  if sni and type(sni) ~= "string" then
+    error("sni must be a string", 2)
+  end
+  if req_headers and type(req_headers) ~= "table" then
+    error("headers must be a table", 2)
+  end
+
+  local c = context.new(self.schema)
+
+  for _, field in ipairs(self.fields) do
+    if field == "http.method" then
+      assert(c:add_value("http.method", req_method))
+
+    elseif field == "http.path" then
+      assert(c:add_value("http.path", req_uri))
+
+    elseif field == "http.host" then
+      assert(c:add_value("http.host", req_host))
+
+    elseif field == "net.protocol" then
+      assert(c:add_value("net.protocol", req_scheme))
+
+    elseif field == "tls.sni" then
+      assert(c:add_value("tls.sni", sni))
+
+    elseif req_headers and field:sub(1, 13) == "http.headers." then
+      local h = field:sub(14)
+      local v = req_headers[h]
+
+      if v then
+        if type(v) == "string" then
+          assert(c:add_value(field, v:lower()))
+
+        else
+          -- TODO: support array of values
+          assert(c:add_value(field, v[1]:lower()))
+        end
+      end
+    end
+  end
+
+  local matched = self.router:execute(c)
+  if not matched then
+    return nil
+  end
+
+  local uuid, matched_path, captures = c:get_result()
+
+  return {
+    route           = self.routes[uuid],
+    service         = self.services[uuid],
+    prefix          = matched_path,
+    matches = {
+      uri_captures = (captures and captures[1]) and captures or nil,
+    }
+  }
+end
+
+
+function _M:exec(ctx)
+  local req_method = get_method()
+  local req_uri = ctx and ctx.request_uri or var.request_uri
+  local req_host = var.http_host
+  local req_scheme = ctx and ctx.scheme or var.scheme
+  local sni = server_name()
+
+  local headers
+  if match_headers then
+    local err
+    headers, err = get_headers(MAX_REQ_HEADERS)
+    if err == "truncated" then
+      log(WARN, "retrieved ", MAX_REQ_HEADERS, " headers for evaluation ",
+                "(max) but request had more; other headers will be ignored")
+    end
+
+    headers["host"] = nil
+  end
+
+  local idx = find(req_uri, "?", 2, true)
+  if idx then
+    req_uri = sub(req_uri, 1, idx - 1)
+  end
+
+  req_uri = normalize(req_uri, true)
+
+  local match_t = self:select(req_method, req_uri, req_host, req_scheme,
+                              sni, headers)
+  if not match_t then
+    return
+  end
+
+  -- debug HTTP request header logic
+  if var.http_kong_debug then
+    local route = match_t.route
+    if route then
+      if route.id then
+        header["Kong-Route-Id"] = route.id
+      end
+
+      if route.name then
+        header["Kong-Route-Name"] = route.name
+      end
+    end
+
+    local service = match_t.service
+    if service then
+      if service.id then
+        header["Kong-Service-Id"] = service.id
```

Comment:
It will overwrite the value assigned above, maybe we can do some optimization?

Reply:
Could you elaborate? This is `header["Kong-Service-Id"]`, different from above which is `Kong-Route-Id`.

Reply:
Sorry, it is my mistake.